tools/clang/lib/SPIRV/SpirvInstruction.cpp

```diff
@@ -662,65 +660,66 @@ SpirvFunctionCall::SpirvFunctionCall(QualType resultType, SourceLocation loc,
-    : SpirvGroupNonUniformOp(IK_GroupNonUniformBinaryOp, op, resultType, loc,
-                             scope),
-      arg1(arg1Inst), arg2(arg2Inst) {
-  assert(op == spv::Op::OpGroupNonUniformBroadcast ||
-         op == spv::Op::OpGroupNonUniformBallotBitExtract ||
-         op == spv::Op::OpGroupNonUniformShuffle ||
-         op == spv::Op::OpGroupNonUniformShuffleXor ||
-         op == spv::Op::OpGroupNonUniformShuffleUp ||
-         op == spv::Op::OpGroupNonUniformShuffleDown ||
-         op == spv::Op::OpGroupNonUniformQuadBroadcast ||
-         op == spv::Op::OpGroupNonUniformQuadSwap);
-}
-
-SpirvNonUniformElect::SpirvNonUniformElect(QualType resultType,
-                                           SourceLocation loc, spv::Scope scope)
-    : SpirvGroupNonUniformOp(IK_GroupNonUniformElect,
-                             spv::Op::OpGroupNonUniformElect, resultType, loc,
-                             scope) {}
-
-SpirvNonUniformUnaryOp::SpirvNonUniformUnaryOp(
-    spv::Op op, QualType resultType, SourceLocation loc, spv::Scope scope,
-    llvm::Optional<spv::GroupOperation> group, SpirvInstruction *argInst)
-    : SpirvGroupNonUniformOp(IK_GroupNonUniformUnaryOp, op, resultType, loc,
-                             scope),
-      arg(argInst), groupOp(group) {
-  assert(op == spv::Op::OpGroupNonUniformAll ||
-         op == spv::Op::OpGroupNonUniformAny ||
-         op == spv::Op::OpGroupNonUniformAllEqual ||
-         op == spv::Op::OpGroupNonUniformBroadcastFirst ||
-         op == spv::Op::OpGroupNonUniformBallot ||
-         op == spv::Op::OpGroupNonUniformInverseBallot ||
-         op == spv::Op::OpGroupNonUniformBallotBitCount ||
-         op == spv::Op::OpGroupNonUniformBallotFindLSB ||
-         op == spv::Op::OpGroupNonUniformBallotFindMSB ||
-         op == spv::Op::OpGroupNonUniformIAdd ||
-         op == spv::Op::OpGroupNonUniformFAdd ||
-         op == spv::Op::OpGroupNonUniformIMul ||
-         op == spv::Op::OpGroupNonUniformFMul ||
-         op == spv::Op::OpGroupNonUniformSMin ||
-         op == spv::Op::OpGroupNonUniformUMin ||
-         op == spv::Op::OpGroupNonUniformFMin ||
-         op == spv::Op::OpGroupNonUniformSMax ||
-         op == spv::Op::OpGroupNonUniformUMax ||
-         op == spv::Op::OpGroupNonUniformFMax ||
-         op == spv::Op::OpGroupNonUniformBitwiseAnd ||
-         op == spv::Op::OpGroupNonUniformBitwiseOr ||
-         op == spv::Op::OpGroupNonUniformBitwiseXor ||
-         op == spv::Op::OpGroupNonUniformLogicalAnd ||
-         op == spv::Op::OpGroupNonUniformLogicalOr ||
-         op == spv::Op::OpGroupNonUniformLogicalXor);
+SpirvGroupNonUniformOp::SpirvGroupNonUniformOp(
+    spv::Op op, QualType resultType, spv::Scope scope,
+    llvm::ArrayRef<SpirvInstruction *> operandsVec, SourceLocation loc,
+    llvm::Optional<spv::GroupOperation> group)
+    : SpirvInstruction(IK_GroupNonUniformOp, op, resultType, loc),
+      execScope(scope), operands(operandsVec.begin(), operandsVec.end()),
```

Comment:
Given that the constructor always copies the operands, would it make more sense to make the parameter a `llvm::SmallVector<SpirvInstruction *, 4>` and use `std::move` to move the args into the `operands` data member?

Actualy, after some thought, I suspect the answer is that it doesn't really matter, since the number of args is expected to be small, so we won't be heap allocating, and thus never benefit from moving. Just leaving this comment here in case you can teach me more about the DXC codebase.

Reply:
Since SmallVector avoids heap allocations and is `std::is_trivially_copyable`, I think you're right that there's no benefit to be gained unless the size is sometimes >4, which it never should be in this case. Some more info here: https://llvm.org/docs/ProgrammersManual.html#llvm-adt-smallvector-h